plugins/janus_audiobridge.c

```diff
@@ -2831,16 +2863,90 @@ void janus_audiobridge_incoming_rtp(janus_plugin_session *handle, int video, cha
 		int plen = 0;
 		const unsigned char *payload = (const unsigned char *)janus_rtp_payload(buf, len, &plen);
 		if(!payload) {
-			JANUS_LOG(LOG_ERR, "[Opus] Ops! got an error accessing the RTP payload\n");
+			JANUS_LOG(LOG_ERR, "(participant %"SCNu64")[Opus] Ops! got an error accessing the RTP payload\n", participant->user_id);
 			g_free(pkt->data);
 			g_free(pkt);
 		  	participant->working = FALSE;
 			return;
 		}
-		pkt->length = opus_decode(participant->decoder, payload, plen, (opus_int16 *)pkt->data, BUFFER_SAMPLES, USE_FEC);
+
+
+		//Check sequence number received, verify if relevant to expected one.
+		if(pkt->seq_number == participant->expected_seq) {
+			/* regular decode */
+            //output_samples = max_frame_size;
+            pkt->length = opus_decode(participant->decoder, payload, plen, (opus_int16 *)pkt->data, BUFFER_SAMPLES, 0);
+
+            //update last_timestamp
+            participant->last_timestamp = pkt->timestamp;
+
+            //Increment according to previous seq_number
+			participant->expected_seq = pkt->seq_number + 1;
+		} else if(pkt->seq_number > participant->expected_seq) {
+			//Sequence(s) losts
+			uint16_t gap = pkt->seq_number - participant->expected_seq;
+			JANUS_LOG(LOG_VERB, "(participant %"SCNu64"),   %"SCNu16" sequence(s) lost, sequence = %"SCNu16",  expected seq = %"SCNu16" \n", participant->user_id, gap, pkt->seq_number, participant->expected_seq);
+
+			//Use FEC if sequence lost < DEFAULT_PREBUFFERING
+			uint16_t start_lost_seq = participant->expected_seq;
+			if(use_fec && gap < DEFAULT_PREBUFFERING) {
+                for(uint8_t i = 1; i <= gap ; i++) {
+					int32_t output_samples;
+					janus_audiobridge_rtp_relay_packet *lost_pkt = g_malloc(sizeof(janus_audiobridge_rtp_relay_packet));
+					lost_pkt->data = g_malloc0(BUFFER_SAMPLES*sizeof(opus_int16));
+					lost_pkt->ssrc = 0;
+					lost_pkt->timestamp = participant->last_timestamp + (i * 960);
+					lost_pkt->seq_number = start_lost_seq++;
+					lost_pkt->silence = FALSE;
+					lost_pkt->length = 0;
+
+					JANUS_LOG(LOG_VERB, "(participant %"SCNu64"), Prepare lost packet with seq = %"SCNu16", timestamp = %"SCNu32" \n", participant->user_id, lost_pkt->seq_number, lost_pkt->timestamp);
+					if(i == gap) {
+						/* attempt to decode with in-band FEC from next packet */
+                		opus_decoder_ctl(participant->decoder, OPUS_GET_LAST_PACKET_DURATION(&output_samples));
+                		JANUS_LOG(LOG_VERB, "(participant %"SCNu64",   FEC(1) output_samples =  %"SCNu32" \n", participant->user_id, output_samples);
+                		lost_pkt->length = opus_decode(participant->decoder, payload, plen, (opus_int16 *)lost_pkt->data, output_samples, 1);
+                	} else {
+                		opus_decoder_ctl(participant->decoder, OPUS_GET_LAST_PACKET_DURATION(&output_samples));
```

Comment:
What's the purpose of this call, here? Isn't this the non-FEC branch?

Reply:
No, here it's the case of audio loss concealment with repair technique associate to the codec (silence substitution ...)